### PR TITLE
Update our contributing document

### PR DESCRIPTION
**Pull request by jaredpar**
Documentation only change

This changes our contributing document in the following ways:

- Moves it from the WIKI to the repo root as CONTRIBUTING.md.  This is the location Github and developers expects to find the file.
- Change it to be very guideline focused.  A lot of it was workflow focused and that is covered elsewhere.  This should be a succinct document helping guide contributions to our repo.
- Removed out of date information.  

Relevant Github posts: https://help.github.com/articles/setting-guidelines-for-repository-contributors/

### CODE REVIEW

**Comment by jaredpar** [comment]
CC @Pilchie, @sharwell, @jasonmalinowski

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+
+For now, the team has set the following limits on pull requests:
+
+- Contributions beyond the level of a bug fix must be discussed with the team first, or they will likely be declined.
+- Pull requests must pass all Jenkins legs and be merge conflict free before they will be merged. 
+- Submissions must meet functional and performance expectations, including scenarios for which the team doesn’t yet have open source tests. This means you may be asked to fix and resubmit your pull request against a new open test case if it fails one of these tests.
+- Submissions must follow the [.editorconfig](http://editorconfig.org/) settings for each directory. 
+- Contributors must sign the [.NET CLA](https://cla2.dotnetfoundation.org/)
+
+When you are ready to proceed with making a change, get set up to [build](https://github.com/dotnet/roslyn/blob/master/docs/contributing/Building%2C%20Debugging%2C%20and%20Testing%20on%20Windows.md) the code and familiarize yourself with our developer workflow.
```
Comment (jasonmalinowski) [inline CONTRIBUTING.md]: Is it possible for this to not reference the master branch directly? Consider making changes in a PR to this, it might be a bit easier to figure out.

File: docs/contributing/Building, Debugging, and Testing on Unix.md [hunk 2]
```diff
@@ -10,4 +10,4 @@ After building run `make test` in order to run the unit tests.
 
 # Contributing
 
-Please see [Contributing Code](https://github.com/dotnet/roslyn/wiki/Contributing-Code) for details on contributing changes back to the code.
+Please see [Contributing Code](https://github.com/dotnet/roslyn/blob/master/CONTRIBUTING.md) for details on contributing changes back to the code.
```
Comment (jasonmalinowski) [inline docs/contributing/Building, Debugging, and Testing on Unix.md]: Ditto to earlier question: is it possible to make this branch-relative so if you're in one branch it doesn't jump between?

File: docs/contributing/Building, Debugging, and Testing on Windows.md [hunk 3]
```diff
@@ -119,4 +119,4 @@ csc and vbc inside it. You can check the cibuild.cmd and see how it is used.
 
 ## Contributing
 
-Please see [Contributing Code](https://github.com/dotnet/roslyn/wiki/Contributing-Code) for details on contributing changes back to the code.
+Please see [Contributing Code](https://github.com/dotnet/roslyn/blob/master/CONTRIBUTING.md)) for details on contributing changes back to the code.
```
Comment (jasonmalinowski) [inline docs/contributing/Building, Debugging, and Testing on Windows.md]: Ditto to earlier question: is it possible to make this branch-relative so if you're in one branch it doesn't jump between?

File: CONTRIBUTING.md (hunk 1 above)
Comment (jaredpar) [inline CONTRIBUTING.md]: No idea.  I've tried using relative links and not had a lot of success.  Perhaps @sharwell knows a way. 

I agree it would be nice.  I just don't know if there is a way to do this.

Comment (sharwell) [inline CONTRIBUTING.md]: In reality this file almost never changes because it's confusing to everyone when it does. I don't think you can do dynamic links here, but I've also never seen it cause a problem. I'd almost go so far as to say if it does cause a problem, we're probably doing something wrong anyway. 😄

**Comment by jaredpar** [comment]
@jasonmalinowski, @sharwell, @Pilchie 

Was browsing around at other CONTRIBUTING.md files this morning.  It seems in predominant style in dotnet is more of a DO / DO NOT list.  Changed this PR to match that style.  Please let me know if you prefer the original.

File: CONTRIBUTING.md [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+
+### CSharp
+
+- **DO** use the coding style outlined in the [.NET Foundation Coding Guidelines](https://github.com/dotnet/corefx/blob/master/Documentation/coding-guidelines/coding-style.md)
+- **DO** use plain code to validate parameters at public boundaries. Do not use Contracts or magic helpers.
+
+```csharp
+if (argument == null)
+{
+    throw new ArgumentNullException("argument");
```
Comment (sharwell) [inline CONTRIBUTING.md]: ❕ `nameof(argument)`

File: CONTRIBUTING.md [hunk 5]
```diff
@@ -0,0 +1,56 @@
+# Contributing to Roslyn
+
+Guidelines for contributing to the Roslyn repo.
+
+## Submitting Pull Requests
+
```
Comment (sharwell) [inline CONTRIBUTING.md]: :bulb: Before having a bunch of scary rules, start with a comment that says we're here to help and invites the user to ask questions if they don't understand anything.

File: CONTRIBUTING.md [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+- **DO** ensure submissions pass all Jenkins legs and are merge conflict free.
+- **DO** ensure submissions meet functional and performance expectations, including scenarios for which the team doesn’t yet have open source tests. This means you may be asked to fix and resubmit your pull request against a new open test case if it fails one of these tests.
```
Comment (sharwell) [inline CONTRIBUTING.md]: Curly quotes make me sad 🙁

File: CONTRIBUTING.md [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+- **DO** use plain code to validate parameters at public boundaries. Do not use Contracts or magic helpers.
+
+```csharp
+if (argument == null)
+{
+    throw new ArgumentNullException("argument");
+}
+```
+
+- **DO** use `Debug.Assert()` for checks not needed in retail builds. Always include a “message” string in your assert to identify failure conditions. Add assertions to document assumptions on non-local program state or parameter values, e.g. “At this point in parsing the scanner should have been advanced to a ‘.’ token by the caller”.
```
Comment (sharwell) [inline CONTRIBUTING.md]: :bulb: `` `message` ``

File: CONTRIBUTING.md [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+- **DO** avoid allocations in compiler hot paths:
+    - Avoid LINQ.
+    - Avoid using foreach over collections that do not have a struct enumerator.
```
Comment (sharwell) [inline CONTRIBUTING.md]: :bulb: `` `foreach` ``

File: CONTRIBUTING.md [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+- **DO** specify a detailed description of the issue or requested feature.
+- **DO** provide the following for bug reports
+    - Describe the expected behavior and the actual behavior. If it is not self-evident such as in the case of a crash, provide an explanation for why the expected behavior is expected.
+    - Provide example code that reproduces the issue.
+    - Specify any relevant exception messages and stack traces.
+- **DO** subscribe to notifications for the created issue in case there are any follow up questions.
+
+## Coding Style
+
+The Roslyn project is a member of the [.NET Foundation](https://github.com/orgs/dotnet) and follow the same [developer guide](https://github.com/dotnet/corefx/blob/master/Documentation/coding-guidelines/coding-style.md).  The repo also includes [.editorconfig](http://editorconfig.org) files to help enforce this convention.  Contributors should ensure they follow these guidelines when making submissions.
```
Comment (sharwell) [inline CONTRIBUTING.md]: :bulb: Mention that most **.editorconfig** features are applied automatically when working in Visual Studio 2017.

File: CONTRIBUTING.md [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+
+## Creating Issues
+
+- **DO** use a descriptive title that identifies the issue to be addressed or the requested feature. For example when describing an issue where the compiler is not behaving as expected, write your bug title in terms of what the compiler should do rather than what it is doing – “C# compiler should report CS1234 when Xyz is used in Abcd.”
+- **DO** specify a detailed description of the issue or requested feature.
+- **DO** provide the following for bug reports
+    - Describe the expected behavior and the actual behavior. If it is not self-evident such as in the case of a crash, provide an explanation for why the expected behavior is expected.
+    - Provide example code that reproduces the issue.
+    - Specify any relevant exception messages and stack traces.
+- **DO** subscribe to notifications for the created issue in case there are any follow up questions.
```
Comment (sharwell) [inline CONTRIBUTING.md]: :memo: This is automatic, so stating it here seems unnecessary.

File: CONTRIBUTING.md [hunk 11]
```diff
@@ -0,0 +1,56 @@
...
+
+Guidelines for contributing to the Roslyn repo.
+
+## Submitting Pull Requests
+
+- **DO** ensure submissions pass all Jenkins legs and are merge conflict free.
+- **DO** ensure submissions meet functional and performance expectations, including scenarios for which the team doesn’t yet have open source tests. This means you may be asked to fix and resubmit your pull request against a new open test case if it fails one of these tests.
+- **DO** follow the [.editorconfig](http://editorconfig.org/) settings for each directory. 
+- **DO NOT** submit changes beyond the level of a bug fix without discussing with the team first, or they will likely be declined.
+- **DO NOT** submit unsolicited language features.  All language feature discussion should begin on the [C# language](https://github.com/dotnet/csharplang#discussion) / [VB language](https://github.com/dotnet/vblang) repos.  Once a feature is championed and validated by LDM a developer will be assigned to help begin a prototype on this repo inside a feature branch.
```
Comment (sharwell) [inline CONTRIBUTING.md]: :bulb: Rephrase this to be a **DO** item.

> **DO** submit a proposal to [dotnet/csharplang]() and/or [dotnet/vblang]() when implementing a new language feature. ...

File: CONTRIBUTING.md [hunk 12]
```diff
@@ -0,0 +1,56 @@
+# Contributing to Roslyn
+
+Guidelines for contributing to the Roslyn repo.
+
+## Submitting Pull Requests
+
+- **DO** ensure submissions pass all Jenkins legs and are merge conflict free.
+- **DO** ensure submissions meet functional and performance expectations, including scenarios for which the team doesn’t yet have open source tests. This means you may be asked to fix and resubmit your pull request against a new open test case if it fails one of these tests.
+- **DO** follow the [.editorconfig](http://editorconfig.org/) settings for each directory. 
+- **DO NOT** submit changes beyond the level of a bug fix without discussing with the team first, or they will likely be declined.
```
Comment (sharwell) [inline CONTRIBUTING.md]: :bulb: Rephrase this to a **DO** item, remove the "or else".

> **DO** submit issues for behavior changes that are not obviously bug fixes. This facilitates discussion of the feature separately from the implementation, and increases the acceptance rates for pull requests.

File: CONTRIBUTING.md (hunk 11 above)
Comment (jasonmalinowski) [inline CONTRIBUTING.md]: Perhaps "prior", but I love the conversion to the positive tone.

File: CONTRIBUTING.md (hunk 6 above)
Comment (jasonmalinowski) [inline CONTRIBUTING.md]: Not sure this one translates well to a "DO" form. We're telling them to ensure it passes things they don't see until we tell them it's broken. And with no link to what those expectations are, we're telling them to follow rules they don't know. Let's consider just deleting this line, I'm not sure what value it brings.

File: CONTRIBUTING.md [hunk 13]
```diff
@@ -0,0 +1,56 @@
...
+- **DO NOT** submit unsolicited language features.  All language feature discussion should begin on the [C# language](https://github.com/dotnet/csharplang#discussion) / [VB language](https://github.com/dotnet/vblang) repos.  Once a feature is championed and validated by LDM a developer will be assigned to help begin a prototype on this repo inside a feature branch.
+- **DO NOT** submit large code formatting changes. Such changes will be declined.
```
Comment (jasonmalinowski) [inline CONTRIBUTING.md]: Given we ourselves in the IDE do this on occasion, this isn't a correct statement. Generally I think the answer is "talk to us in advance and make a separate PR".

File: CONTRIBUTING.md (hunk 10 above)
Comment (jaredpar) [inline CONTRIBUTING.md]: You'd be surprised 😦

File: CONTRIBUTING.md (hunk 6 above)
Comment (jaredpar) [inline CONTRIBUTING.md]: Think that was done by @Pilchie and I just copied it.

Comment (jaredpar) [inline CONTRIBUTING.md]: Glad someone else felt that way too.  This is a hold over from our original attempt at documenting our processes.  Given the very low number of times we've had to do this I think it's fine to remove.

File: CONTRIBUTING.md (hunk 11 above)
Comment (jaredpar) [inline CONTRIBUTING.md]: Generally I understand the desire to be more positive here.  However I think we need a very direct statement of what will happen when we get PRs that don't meet our needs.  In particular language features.  Lacking a statement on our repo that we close unsolicited language changes it ends up falling to me to say it to each individual developer.  I think it's better to be up front about this.

File: CONTRIBUTING.md (hunk 13 above)
Comment (jaredpar) [inline CONTRIBUTING.md]: I'll add unsolicited.

**Comment by jaredpar** [comment]
retest windows_build_correctness_prtest please

**Comment by jaredpar** [comment]
The correctness build failed due to NuGet sync issues.
